File: Ejercicio1/src/Carrera.java

import java.util.ArrayList;

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author juanfrandm98
 */
public abstract class Carrera extends Thread {
    
    private ArrayList<Bicicleta> bicicletas;
    private TipoBicicleta tipo;
    private int porcentajeAbandono;
    
    public Carrera( TipoBicicleta tipo, int porcentajeAbandono ) {
        bicicletas = new ArrayList();
        this.tipo = tipo;
        this.porcentajeAbandono = porcentajeAbandono;
    }
    
    public void aniadirBicicleta( Bicicleta bicicleta ) {
        
        int nuevoIdentificador = bicicletas.size() - 1;
        bicicleta.setIdentificador( nuevoIdentificador );
        
        bicicletas.add( bicicleta );
        
    }
    public int abandonarCarrera(){
        int abandonan = (this.porcentajeAbandono * bicicletas.size())/100;
        for(int i = 0 ; i < abandonan ; i++){
            bicicletas.remove(i);
        }
        return abandonan;
    }
    
    
    
}

File: Ejercicio2/Ejercicio2/lib/Carrera.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.


class Carrera
  def initialize(tipo, porcentajeAbandono)
    
      @bicicletas = []
      @tipo = tipo
      @pocentajeAbandono = porcentajeAbandono
    
  end
  
  def self.new()
    raise "Error:Clase Abstracta"
  end
  
  def aniadirbicicleta (bicicleta)
    id = @bicicletas.size
    bicicleta.setidentificador(id)
    
    bicicletas.push(bicicleta)
  end
  
  def abandonarcarrera
    abandonan = (@porcentajeAbandono * @bicicletas.size/100)
    
    abandonan.times do |num|
      siguiente = (rand(@biciletas.size))
      @bicicletas.delete_at(siguiente)
    end
    
    return abandonan
  end
  
  def clone
    raise "Abstracto"
  end
  
  def run
    mensaje = "Iniciando carrera de #{@tipo} con bicicletas"
    
    @bicicletas.size.times do |num|
      mensaje += " #{@bicicletas[num].getidentificador}"
    end
    
    puts mensaje
    
    sleep(5)
    numAbandonos = abandonarcarrera()
    
    puts "Carrera #{@tipo} - han abandonando #{numAbandonos} bicicletas"
    
    sleep(5)
    
    mensaje = "Carrera de #{@tipo} terminada con biciletas: "
    
     @bicicletas.size.times do |num|
      mensaje += " #{@bicicletas[num].getidentificador}"
    end
    
    puts mensaje
  end
end